### PR TITLE
feat: Schema for the oldstyle edm4hep Future Circular Collider simulation Samples

### PR DESCRIPTION
**Pull request by prayagyadav**
Hi,
This is a schema for the Future Circular Collider Spring2021 pre-generated samples based on oldstyle EDM4HEP. This is a work in progress and I wanted to post it here for community inputs.
I have added most of the features: 

- [x] Added mixing classes most of the EDM4HEP classes
- [x] Association between ReconstructedParticle and MCParticle
- [x] Methods to get the daughters and parents to each MCParticle 
- [ ] Special methods for Tracks and Clusters (Need input from the community)
- [x] Tests and test root file for the FCCSchema

Further development will need inputs from more people.
Thanks

Links and References:
- Used `p8_ee_ZH_ecm240` sample from [here](https://fcc-physics-events.web.cern.ch/FCCee/delphes/spring2021/idea/) for testing (generated a smaller file with 100 events)
- [EDM4HEP](https://edm4hep.web.cern.ch/)
- [Structure of EDM4HEP files](https://github.com/HEP-FCC/FCCAnalyses/tree/master/examples/basics#structure-of-edm4hep-files)

Tagging my guides: @davidlange6 @gomber

### CODE REVIEW

**Comment by lgray** [comment]
Hi! Sorry for taking some time to get back to you on this.

This schema looks really well put together, thank you, and thanks for adding some basic tests.

I'll get through a review soon.

**Comment by lgray** [comment]
It's a bit less clean than yours but you may also want to have a look at https://github.com/CoffeaTeam/coffea/pull/822. It's more a work in progress though.

**Comment by lgray** [comment]
What would be everyone's opinion on generalizing this to any edm4hep? It should be relatively easy, mostly you have to deal with missing branches as in NanoAOD.

**Comment by prayagyadav** [comment]
@lgray Thanks for taking the time to review. #822 was indeed, a starting point for the development of FCCSchema. Many pre-generated samples in [FCC Events](https://fcc-physics-events.web.cern.ch/), especially the [Spring2021](https://fcc-physics-events.web.cern.ch/FCCee/delphes/spring2021/idea/) and [Winter2023](https://fcc-physics-events.web.cern.ch/FCCee/delphes/winter2023/idea/), are `pre-edm4hep1`(i.e. the oldstyle EDM4HEP I believe).

Yeah, creating a similar schema for the FCCSamples based on new style edm4hep ([eg. file](https://cernbox.cern.ch/s/s5C1ohEj1mWIzoy) generated by @davidlange6) must be fairly easy.  

Looks like #822 describes a general edm4hep schema, but the names of the branches assumed there, do not match with ones in FCC events. It would be really good if we could have a general edm4hep schema, and fcc schema could inherit from that, adding fcc specific functionality.

Nevertheless, the current FCCSchema( oldstyle edm4hep ) needs to be there, to facilitate interpretation of the older campaigns (Spring2021, Winter2023, etc.)

@davidlange6 What's your opinion on this?

**Comment by lgray** [comment]
Oh those were some initial comments. I have more that I want to go through, but it'll be monday or so.

**Comment by davidlange6** [comment]
> What would be everyone's opinion on generalizing this to any edm4hep? It should be relatively easy, mostly you have to deal with missing branches as in NanoAOD.

right - its the eventual goal.. @prayagyadav and I thought to use this as an initial version to get feedback (and eventually users) on

File: src/coffea/nanoevents/methods/fcc.py [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+        # operating on length-zero data returns the wrong layout!
+        # We need the axis 2, therefore, we should return the typetracer layout of [[[]]]
+        awkward.typetracer.length_zero_if_typetracer(
+            begin
+        )  # force touching of the necessary data
+        awkward.typetracer.length_zero_if_typetracer(
+            end
+        )  # force touching of the necessary data
+        return awkward.Array(
+            awkward.Array([[[]]]).layout.to_typetracer(forget_length=True)
```
Comment (lgray) [inline src/coffea/nanoevents/methods/fcc.py]: This return value for the TypeTracer does not preserve the contained data type. This could lead to confusion.

It would be better to do something like:

```python3
if ak.backend(begin_end) == "typetracer":
     return index_range_numba_wrap(
         awkward.typetracer.length_one_if_typetracer(begin_end),
         awkward.ArrayBuilder(),
     ).snapshot().layout.to_typetracer(forget_length=True)
```

Comment (prayagyadav) [reply src/coffea/nanoevents/methods/fcc.py]: Okayy. Will check it

File: src/coffea/nanoevents/methods/fcc.py [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+    @get_daughters.dask
+    def get_daughters(self, dask_array):
+        """
+        Obtain the actual MCParticle daughters to each and every MCParticle
+        - The output is a doubly nested awkward array
+        - Needs the presence of Particleidx1 collection
+        - The Particleidx1.index contains info about the daughters
+        """
+        return map_index_to_array(dask_array, dask_array.get_daughters_index, axis=2)
```
Comment (lgray) [inline src/coffea/nanoevents/methods/fcc.py]: For this function and others below it you'll likely have the same problem we did with the NanoAOD schema where this will not work after cuts are placed on the object, since the indices are local to each event.

The solution we came up with was to reference back to the original array and then take only the objects that passed the original cuts. You can see an example for that flavor of GenParticles here: https://github.com/CoffeaTeam/coffea/blob/master/src/coffea/nanoevents/methods/nanoaod.py#L119

The main thing you have to do is generate the global indices upon reading in the file so that you can use those arrays instead of the local indices.

File: src/coffea/nanoevents/methods/fcc.py [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    @dask_property
+    def matched_gen(self):
+        sel = awkward.broadcast_arrays(True, self)[0]
+        index = self._events().MCRecoAssociations.reco_mc_index[:, :, 1]
+        return self._events().Particle[index[sel]]
+
+    @matched_gen.dask
+    def matched_gen(self, dask_array):
+        sel = awkward.broadcast_arrays(True, dask_array)[0]
+        index = dask_array._events().MCRecoAssociations.reco_mc_index[:, :, 1]
```
Comment (lgray) [inline src/coffea/nanoevents/methods/fcc.py]: Same issue with the indexing through cuts. I had most of an example together in #822.

File: src/coffea/nanoevents/methods/fcc.py (hunk 2 above)
Comment (prayagyadav) [reply src/coffea/nanoevents/methods/fcc.py]: Hi @lgray. Sorry for the super late reply. Indeed, this won't work if cuts are already applied. But I have trouble understanding the solution you suggested. Furthermore, I don't know what `_apply_global_index` actually does. Please explain in simpler terms how I can fix this . Thanks

File: src/coffea/nanoevents/methods/fcc.py (hunk 1 above)
Comment (prayagyadav) [reply src/coffea/nanoevents/methods/fcc.py]: Hi, sorry for the late update. This snippet does not work. I don't really understand the logic. Could you please explain what `length_one_if_typetracer` does? Alternatively, is there a documentation/example use case for this function that I can refer to? Thanks

**Comment by lgray** [comment]
@tmadlener FYI this is was I was talking about yesterday.

**Comment by prayagyadav** [comment]
Hi! @lgray ,  I figured out how to add global indexers and cross-references. I think I have fixed all of the issues that you had pointed out.  Sorry for the unsquashed commits (I thought I did squash them, but evidently, it did not work). I think I will add some more tests later to check the validity of parents and daughters cross references. For now, I have this simple test in [this notebook](https://github.com/prayagyadav/coffea/blob/master/dev/Parents_Daughters_Global_Indexer_Test.ipynb).
Please let me know what do you think about the changes. Thank you.

**Comment by lgray** [comment]
@prayagyadav Thanks for following this up despite lack of feedback. I was busy at a workshop. What you have implemented looks correct.

I'll wait for your additional tests and then we'll merge!

**Comment by lgray** [comment]
Closes #822

File: src/coffea/nanoevents/transforms.py [hunk 4]
```diff
@@ -141,6 +141,62 @@ def local2global(stack):
...
+    and
+    target is an example target branch
+    '''
+    target = stack.pop()
+    index = stack.pop()
+
+    counts1 = awkward.num(target, axis=1)
+    counts2 = awkward.num(index, axis=1) # This is, in reality, flattened counts for axis = 2
+    
+    print(target)
```
Comment (lgray) [inline src/coffea/nanoevents/transforms.py]: please remove these print statements

Comment (lgray) [reply src/coffea/nanoevents/transforms.py]: oops, that was an old commit.

**Comment by prayagyadav** [comment]
Hi @lgray, I have added the tests for parents and daughters. I have also added one more sample file (needed to test the winter2023 campaign of FCC-Events).

**Comment by lgray** [comment]
@prayagyadav For your 4th checkbox, do you want to address that in a later PR, or are you waiting for some more prompt feedback?

**Comment by prayagyadav** [comment]
> @prayagyadav For your 4th checkbox, do you want to address that in a later PR, or are you waiting for some more prompt feedback?

I can address that in a later PR when I figure out that part of the functionality.

**Comment by lgray** [comment]
@prayagyadav all looks good - auto-merge enabled on this latest CI run. Thanks a lot for your work!

We should plan with @davidlange6 and @tmadlener about extending this to new-style edm4hep.

I have made a tracking issue for the Tracks and Clusters functionality. We'll follow up in time.

**Comment by tmadlener** [comment]
For completeness, and because it might be useful for development: [This workflow of EDM4hep](https://github.com/key4hep/EDM4hep/actions/workflows/key4hep-build.yaml) creates a dummy EDM4hep file that contains all available datatypes and relations between them. That should give you all the branch names (and actual c++ datatypes if that matters here) for EDM4hep.

**Comment by prayagyadav** [comment]
> @prayagyadav all looks good - auto-merge enabled on this latest CI run. Thanks a lot for your work!
> 
> We should plan with @davidlange6 and @tmadlener about extending this to new-style edm4hep.
> 
> I have made a tracking issue for the Tracks and Clusters functionality. We'll follow up in time.

Awesome!
